File: compilers/python/femto.py
import ast
import meta
import inspect
import femto_pb2 as schema

nil = schema.Expression()
nil.nil = True

def sym(symbol):
    e = schema.Expression()
    e.symbol = symbol
    return e

def s(value):
    e = schema.Expression()
    e.string = value
    return e

def n(value):
    e = schema.Expression()
    e.number = float(value)
    return e

def l(*items):
    e = schema.Expression()
    e.list.items.extend(items)
    return e

def function(arguments, body):
    e = schema.Expression()
    e.function.arguments.extend(arguments)
    e.function.body.CopyFrom(body)
    return e

def cond(predicate, truth, falsehood):
    e = schema.Expression()
    e.condition.predicate.CopyFrom(predicate)
    e.condition.truth.CopyFrom(truth)
    e.condition.falsehood.CopyFrom(falsehood)
    return e
    
def bind(symbol, expression):
    b = schema.Binding()
    b.symbol = symbol.symbol
    b.expression.CopyFrom(expression)
    return b

def let(bindings, body):
    e = schema.Expression()
    for binding in bindings:
        b = e.let.bindings.add()
        b.CopyFrom(binding)
    e.let.body.CopyFrom(body)

    return e

def apply(operation, arguments):
    e = schema.Expression()
    e.apply.operation.CopyFrom(operation)
    e.apply.arguments.extend(arguments)
    return e

### (function [x]
###   (reduce
###    (function [a b] (+ a b))
###    0 (vals (get x "count"))))

# addTree = function(['a', 'b'],
#              apply(sym('+'), [sym('a'), sym('b')]))

# foldTree = function(['x'],
#               apply(sym('reduce'),
#                     [addTree,
#                      n(0),
#                      apply(sym('vals'),
#                            [apply(sym('get'),
#                                   [sym('x'),
#                                    s('count')])])]))

# print foldTree

def nativeSum(x):
    counts = x['count'].vals()
    return reduce(lambda a, b: a + b, 0, counts)

def tamp(z):
    return [x for y in z for x in y]
    
def astForDef(d):
    source = inspect.getsource(d)
    tree = ast.parse(source)
    return tree

def compileFemto(token):
    product = nil

    if isinstance(token, str):
        product = s(token)

    elif isinstance(token, ast.Name):
        product = sym(token.id)

    elif isinstance(token, ast.Num):
        product = n(token.n)

    elif isinstance(token, ast.Str):
        product = s(token.s)

    elif isinstance(token, ast.Add):
        product = sym('+')

    elif isinstance(token, ast.Index):
        product = compileFemto(token.value)

    elif isinstance(token, ast.Subscript):
        get = s('get')
        m = compileFemto(token.value)
        key = compileFemto(token.slice)
        product = apply(get, [m, key])

    elif isinstance(token, ast.Attribute):
        get = s('get')
        m = compileFemto(token.value)
        key = compileFemto(token.attr)
        product = apply(get, [m, key])

    elif isinstance(token, ast.Assign):
        target = compileFemto(token.targets[0])
        value = compileFemto(token.value)
        bindings = [bind(target, value)]
        product = let(bindings, nil)

    elif isinstance(token, ast.Call):
        fn = compileFemto(token.func)
        arguments = map(compileFemto, token.args)
        product = apply(fn, arguments)

    elif isinstance(token, ast.Lambda):
        arguments = map(lambda arg: compileFemto(arg).symbol, token.args.args)
        body = compileFemto(token.body)
        product = function(arguments, body)

    elif isinstance(token, ast.BinOp):
        fn = compileFemto(token.op)
        right = compileFemto(token.right)
        left = compileFemto(token.left)
        product = apply(fn, [right, left])

    elif isinstance(token, ast.Return):
        product = compileFemto(token.value)

    elif isinstance(token, ast.FunctionDef):
        arguments = map(lambda arg: compileFemto(arg).symbol, token.args.args)
        body = map(compileFemto, token.body)
        assigns = filter(lambda step: step.HasField("let"), body)
        bindings = tamp(map(lambda l: l.let.bindings, assigns))
        value = filter(lambda step: not step.HasField("let"), body)[-1]
        l = let(bindings, value)
        product = function(arguments, l)

    elif isinstance(token, ast.Module):
        product = compileFemto(token.body[0])

    else:
        print 'never seen %s type before.... compilation will be partial until this type is supported' % token.__class__

    return product

def femto(target):
    tree = astForDef(target)
    compileFemto(tree)

def run():
    print(compileFemto(astForDef(nativeSum)))

# defold = meta.decompile(nativeFold)
# source = meta.dump_python_source(defold)

File: compilers/js/femto.js
function isFunction(f) {
  return !!(f && f.constructor && f.call && f.apply);
}

function isNumber(n) {
  return !isNaN(n) && typeof(n) === 'number'
}

var _bind = Function.prototype.apply.bind(Function.prototype.bind);
Object.defineProperty(Function.prototype, 'bind', {
  value: function(obj) {
    var boundFunction = _bind(this, arguments);
    boundFunction.boundObject = obj;
    return boundFunction;
  }
});

femto = function() {
  function lookup(env, key) {
    if (env.length > 0) {
      var keys = Object.keys(env[0])
      if (keys.indexOf(key) > -1) {
        var found = env[0][key]
        return found
      } else {
        return lookup(env.slice(1), key)
      }
    } else {
      console.log('unbound symbol! ' + key)
    }
  }

  function evaluate(env, femto) {
    var key = Object.keys(femto)[0]
    console.log(key)
    if (key === 'nil') {
      value = null
    } else if (key === 'string' || key === 'number') {
      value = femto[key]
    } else if (key === 'symbol') {
      var value = lookup(env, femto[key])
    } else if (key === 'list') {
      value = femto['list']['items'].map(function(sub) {return evaluate(env, sub)})
    } else if (key === 'condition') {
      var condition = femto['condition']
      var predicate = evaluate(env, condition['predicate'])
      if (predicate === null || predicate == false || predicate == undefined) {
        value = evaluate(env, condition['falsehood'])
      } else {
        value = evaluate(env, condition['truth'])
      }
    } else if (key === 'function') {
      var fn = femto['function']
      var args = fn['arguments']
      var body = fn['body']
      var program = function() {
        var bindings = Array.prototype.slice.call(arguments);
        var scope = args.reduce(function(scope, symbol, index) {
          var found = bindings[index]
          scope[symbol] = found
          return scope
        }, {})
        var newenv = [scope].concat(env)
        var value = evaluate(newenv, body)
        return value
      }
      value = program
    } else if (key === 'let') {
      var assign = femto['let']
      var scope = assign['bindings'].reduce(function(scope, binding) {
        var key = binding['symbol']
        var value = evaluate(env, binding['expression'])
        scope[key] = value
        return scope
      }, {})
      var newenv = [scope].concat(env)
      var value = evaluate(newenv, assign['body'])
      value = value
    } else if (key === 'apply') {
      var apply = femto['apply']
      var op = evaluate(env, apply['operation'])
      var args = apply['arguments'] ? apply['arguments'].map(function(sub) {return evaluate(env, sub)}) : []
      var result = op.apply(op.boundObject, args)
      value = result
    } else {
      console.log('unknown syntax! ' + JSON.stringify(femto))
      value = undefined
    }

    console.log(value)
    return value
  }

  return evaluate
} ();

python = { "function": { "body": { "let": { "body": { "apply": { "operation": { "symbol": "reduce" }, "arguments": [ { "function": { "body": { "apply": { "operation": { "symbol": "+" }, "arguments": [ { "symbol": "b" }, { "symbol": "a" } ] } }, "arguments": [ "a", "b" ] } }, { "number": 0.0 }, { "symbol": "counts" } ] } }, "bindings": [ { "symbol": "counts", "expression": { "apply": { "operation": { "apply": { "operation": { "symbol": "attribute" }, "arguments": [ { "apply": { "operation": { "symbol": "get" }, "arguments": [ { "symbol": "x" }, { "string": "count" } ] } }, { "string": "vals" } ] } } } } } ] } }, "arguments": [ "x" ] }}

environment = [{
  '+': function(a, b) {
    if (isNumber(a)) {
      if (isNumber(b)) {
        return a + b
      } else {
        return a
      }
    } else {
      return isNumber(b) ? b : 0
    }
  },

  'reduce': function(f, i, l) {return l.reduce(f, i)},
  'get': function(m, k) {return m[k]},
  'attribute': function(m, k) {
    var value = m[k]
    if (isFunction(value)) {
      var fn = value.bind(m)
    }
    return fn
  }
}]

var py = femto(environment, python)
var result = py({count: {a: 1, b: 2, c: 3, d: 4, e: 5, f: 6, vals: function(m) {return Object.values(m || this)}}})
console.log(result)
